### DISCUSSION TITLE
How to set XamlRoot for ContentDialog when using C# Markup + MVUX

### DISCUSSION THREAD

**Discussion by liziyu0714** [unoplatform/uno#18598] [Q&A] [answered]
I want to open a ContentDialog through my Model,as the document said,I need to set the Xamlroot.
But I am using the C# Markup + MVUX and don't know how to access the Page .
What should I do?

**Comment by morning4coffe-dev** [comment, marked as answer]
Hey @liziyu0714,

Thanks for your question! I recommend creating a custom service that you initialize with a `XamlRoot` (e.g., from `App.cs` during startup). Once the app loads, this service can provide the value whenever you call it. For more details on opening a `ContentDialog` from a ViewModel, check out this guide: https://platform.uno/docs/articles/features/dialogs.html?tabs=mvux%2Ccs-markup#opening-a-contentdialog-from-a-modelviewmodel 
Let me know if this helps.

**Reply by DevTKSS** [reply]
May you could include a c# ui markup sample? the one below uses xaml markup. c# markup do not give a loaded event by default and I have no Idea how to do that in this case.
thank you in advance

**Reply by morning4coffe-dev** [reply]
Hey @DevTKSS,

It should be fairly similar to what you would do with XAML; you just add an event handler for the Loaded event. You can take an inspiration from the snippet here:
```csharp
    public MainPage()
    {
        this.Loaded += MainPage_Loaded;

        this
            .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
            .Content(new StackPanel()
            .VerticalAlignment(VerticalAlignment.Center)
            .HorizontalAlignment(HorizontalAlignment.Center)
            .Children(
                new TextBlock()
                    .Text("Hello Uno Platform!")
            ));
    }

    private void MainPage_Loaded(object sender, RoutedEventArgs e)
    {
        // Your code
    }
```
I hope this helps.

**Reply by DevTKSS** [reply]
Okay... But please correct me if I am wrong!
In the template, which does the wizard create when choosing not just the blank preset then the recommended, the ui csharp file is not including any other class or method than the mainPage : Page with the nested ui elements. So I understood it this way, that there are is no events there, no other method than this ui Definition, which makes it hard for me to transfer the examples they use with routedEventArgs to the Public async Task thing which do not take the xamlRoot Parameter if I try to provide it when calling the task. Ironicly giving it the page.XamlRoot, it thows a error, that I can not cast MainViewModel (while i am curious that it do not name the type MainView what it is in the forst lines) into xamlRoot🤔 so for me, its hard to understand how a property which is type of xamlRoot can suddenly be an other type at compiling time. So I was searching for a maybe full example that tells me how to do it with this csharp markup and mvux... Hope its understandable😅🙈

**Reply by DevTKSS** [reply]
@morning4coffe-dev don't know if without linking its not notification you or you dont like to answer... could understand that if its simple for a pro.
So you told, to "simply" add a void below which do the initialization of the class you provided in zip.
Tryed to do what you said in the last snipped and stumbling over a question, you might could answer?
Maybe I missunderstood also the mvux thing in connection to the Model (not ViewModel which is generated in c# markup):
In the template Sub "GoToSecond" is called via a simple "() => GoToSecond" which task is placed in the model and uses the navigator extension. Your example is using the MainPage, not MainModel like this. The xaml ui examples are also using the (Object sender, RoutedEventArgs e) which made me think, this seems to not made for c# markup with mvux. Is that thought right, or would you, as more experienced Developer, tell that this is wrong understood? Is a mvux c# app still using this eventbased thing?
I am referring for this to uno documentation Reference/UI/Features/Dialogs:
"[MVUX](https://platform.uno/docs/articles/features/dialogs.html?tabs=mvux%2Ccs-markup#tabpanel_1_mvux)
If you are using the MVUX pattern, you don't need to specify the Command. Instead, you can directly call the dialog display method in your view, it is generated automatically in BindableModel:

C#


public async Task DisplayDeleteFileDialogCommand()
{
    // ...
}"
assuming from this, I thought I will have to use a task in the vm.SomeContentDialog to correctly do that, but if I do so, I see no chance to get on the xamlRoot. For me it seems like a deadlock (maybe just in my head) how to get the xaml root available outside of the page itself. In the previous mentioned Dialog reference is also mentioned to use this RootProvider thing, but "At App initialization",
so I went to the app.xaml.cs file in my application and stumbled over the configureServices where I could addSingleton. But seems like its not satisfied with that and this seems to do no new method on the way, but needs a interface, so the service has to be instance able...
"                .ConfigureServices((context, services) =>
                {
                    // TODO: Register your services
                    //services.AddSingleton<IMyService, MyService>();
                })"
but this would need a Interface, which is not possible while using the void and propertys as static.
Could you give me your Opinion to that? The idea would be, to init it once, but... where? on the MainPage? fine with me, but then we will be stuck again with the fact, that its not application wide, just in mainPage. Its like a dog biting its own tail.

**Reply by liziyu0714** [reply]
@DevTKSS 
> Is a mvux c# app still using this eventbased thing?

You can refer to the document: [Commands](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Mvux/Advanced/[Commands](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Mvux/Advanced/Commands.html).html).
In mvux ,the viewmodel is auto generated from the model, and your methods are packed up automatically to commands so that you can use something like "() => GoToSecond", it has nothing to do with event.

>For me it seems like a deadlock (maybe just in my head) how to get the xaml root available outside of the page itself.

In mvux , you can still use events, and use them in the Page Class (eg. MainPage.cs).
That is to say , you can call Initialize in the Loaded event of the Page:
```cs
public sealed partial class MainPage : Page
{
    public MainPage()
    {
        this.DataContext<MainViewModel>((page, vm) => page
            .NavigationCacheMode(NavigationCacheMode.Required)
            .Background(Theme.Brushes.Background.Default)
            .Content(...);
        this.Loaded += (sender, args) => { XamlRootService.Initialize(XamlRoot); };
    }
}
```


If you still have any questions , please ask me.

**Reply by DevTKSS** [reply]
@liziyu0714 
thanks for reply.
So you got it working?
I am still struggeling to get the dialog shown sadly. Along documentation it should be possible to call this from Model after filling the xamlRoot to that service he gave in the zip file and filling the xamlroot in the loaded event. But (if you maybe seen my opened Question topic) on my end there is nothing showing up and instead the syncronizationContext seems to be suddenly null after going to the model to run the task instead of doing it in the Main Page. Does this happen on your side also using that service? It dont make any sense to me or the others (discussing also on discord) why that do happen.

**Reply by morning4coffe-dev** [reply]
Hey @DevTKSS, sorry for the delay in my response, did you manage to get it working with C#M+MVUX? I can take a look into creating a sample for this as well, if you need one, but similar concepts should apply.

**Reply by DevTKSS** [reply]
sadly no, but changed to xaml markup and the call of the dialog from the xaml.cs codebehind which worked. Like @liziyu0714  said in the comments below, it happend on her side also and I checked with @mikernet via discord also before opening a seperate issue to that. Would appreciate if someone has a working solution to that issue to be able in the future to call a contentdialog from viewmodel again in the future. Issue number to that with full description and provided repo: #18657 
Feel free to also take a look into that. Used the service you provided which is at least working, but got some UI thread problems which makes it not possible for me to get it working from MainModel which should be possible if we look in the documentation which tells its possible from ViewModel with such a service. Seems like there is a problem in the generated code for me, but I am clearly no professional on this.

**Comment by liziyu0714** [comment]
Thanks!
Could you provide an example?
I don't know how to implement the service.

**Reply by morning4coffe-dev** [reply]
Here’s a simple sample I created that you can download and review: [XamlRootSample.zip](https://github.com/user-attachments/files/17533028/XamlRootSample.zip). It uses a basic architecture, making it applicable for both MVVM and MVUX since it relies on static services for simplicity, which is suitable for this case. For your project, you can take the `ShowDialogAsync` method from `MainPage` and move it to your Model/ViewModel—just remember to call `XamlRootService.Initialize` at some point during app startup (you can use the same `Loaded` method with C#M).

**Comment by liziyu0714** [comment]
Yes.I write an similar service to get xamlroot, and it works in MVVM template.
But in my project using mvux it doesn't work, just like yours. And I find that if you call showasync in loaded method(in mvux),it can work.
________________________________

**Reply by DevTKSS** [reply]
but... just a thought: If i can only call it from the loaded event method, I can not trigger it from a button command, so would that make any sense? Possibly I just misunderstood it

**Comment by liziyu0714** [comment]
in mvvm template ,it can be called from model.But in mvux,it can't
________________________________

**Reply by DevTKSS** [reply]
added that point as link to my question topic post, to be maybe an issue. Dont know if they can escalate it to one but I think its relevant that it only happens in mvux
I did try on xaml.cs along the video and suddenly a ContentDialog appeared. But in my opinion its not meant to be that way if in the documentation they telling its able to call it from the model also
